### ISSUE TITLE
[BUG] How to improve the training perfomance in MLX compare to pytorch and keras ?

### ISSUE THREAD

**Issue by thegodone** [ml-explore/mlx#1542]
**Describe the bug**
I have a major issue that I have seen in lot of the cases on other trial. The MLX training gives rarely a good performance while for torch and keras it is more stable and better. This is really a bottleneck to use MLX, as you need to train 10 to 20 time your model to get a good result while torch and keras are systematically in a good range (rmse : 0.50-0.55).
important: models (tf/keras, torch and mlx) have the same number of trainable parameters, and we use the same train, val and test split for the 3 methods).

**To Reproduce**

run several time the following code the best result is jumping out of the pytorch and tf/keras results
https://github.com/thegodone/apple_ai_model/blob/main/AttFP_mlx_faster.ipynb
https://github.com/thegodone/apple_ai_model/blob/main/AttFP_torch.ipynb
https://github.com/thegodone/apple_ai_model/blob/main/AttFP_tf.ipynb

**Expected behavior**
I don't know if it is weights initialization or optimizer that can cause this huge difference between the 3 packages.

**Desktop (please complete the following information):**
see #1531

**Comment by awni** [comment]
- Did you try the suggestion in #1153 to use a linear warmup? 
- Did you try the bias corrected Adam to see if it improves the convergence rate?

If you can try either of those and report back, that would be useful to know.

> Is it possible to add a parameter for Adam optimizer to be strictly identical to pytorch / tensorflow ?

It's possible..would be good to know if it fixes your issue first though.

**Comment by thegodone** [comment]
here the differences between the 3 trainings: 
- train_loss is also different I need to check that 
- LR are not identical that may explain the difference
![comparing](https://github.com/user-attachments/assets/c086a10f-826b-49e1-b05c-39b29b320aed)

one remark, now have similar speed between torch and mlx

**Comment by awni** [comment]
The 0 training loss in MLX seems incorrect particularly given the training MSE seems reasonable. I would double check you are averaging the loss in MLX correctly.

Otherwise it mostly looks reasonable..

**Comment by thegodone** [comment]
I will have a synchronize LR scheduler tomorrow to be sure this is not the part that affect the model deviation. Yes I like at the loss that is strange.

**Comment by thegodone** [comment]
ok I fix the loss and now I have this type of results interestingly pytorch is more efficient than mlx or tensorflow 
![comparing (1)](https://github.com/user-attachments/assets/301567b5-5d47-44e9-aeb4-1bf76a35b7a1)

**Comment by thegodone** [comment]
I use this LR scheduler now in mlx, one potential issue is that in pytorch/tensorflow the LR scheduler is per epoch while in mlx it is per step, is it possible to have an epoch equivalent ?

````
def cosineannealingwarmrestartfactor_(initial_lr, restart, decay_steps, warmup_factor, Tmin):
    schedules = []
    boundaries = []  # Boundaries should be one less than schedules
    base_lr = initial_lr
    schedules.append(optim.cosine_decay(initial_lr, decay_steps, Tmin))
    for i in range(restart-1):        
        Tmin *= warmup_factor
        initial_lr*=warmup_factor

        schedules.append(optim.cosine_decay(initial_lr, decay_steps, Tmin))
        boundaries.append(decay_steps*(i+1))
        
    lr_schedule = optim.join_schedules(schedules, boundaries)
    return lr_schedule


````

**Comment by thegodone** [comment]
this is the mlx version with "official Adam without bias correction": clearly not performing as the bias correction used in previous posts.
![comparing (2)](https://github.com/user-attachments/assets/19d6bd6f-60a0-4aeb-a9e9-dc70a49c19f9)

**Comment by thegodone** [comment]
Yes, was expected that was the case.The good accuracy on torch is due to a specific seed… so we can close that issue.Envoyé de mon iPhoneLe 12 nov. 2024 à 15:21, Awni Hannun ***@***.***> a écrit :﻿
mx.random.seed

**Comment by angeloskath** [comment]
Closing the issue per the OP's last message ie the difference is due to a specific random seed in PT.
